### ISSUE TITLE
Don't allow pulling all tags by default for the ImageCreate API endpoint

### ISSUE THREAD

**Issue by adibsaad** [moby/moby#42053]
Hi,

I was debugging an issue a while back wherein a program calling the Create Image API endpoint ([link](https://docs.docker.com/engine/api/v1.41/#operation/ImageCreate)) was taking a long time to complete. I realized, after debugging, that when the tag parameter is not specified (which the program wasn't), it pulls all tags.

This is problematic as Docker introduced API limits and people who may not be aware of this behavior may download more than what they probably wanted and will use up their quotas, which is what happened a few times in my case before I realized what was happening.

Since there are cases where people will want to pull all tags (e.g. private registry), I suggest we keep this functionality but make the following changes

- If a user wants to pull all tags, pass in the wildcard character (`*`) as the tag parameter
- If the tag parameter is empty, then only pull the `latest` tag

Thank you

**Comment by thaJeztah** [comment]
Thanks for reporting; yes, this is definitely confusing. I not too long ago changed a similar behaviour for `docker push` (https://github.com/moby/moby/pull/40302), and described the background for this behavior in https://github.com/docker/cli/issues/2214

> The `docker push` command up until [v0.9.1](https://github.com/moby/moby/blob/v0.9.1/api/client.go#L998) always pushed all tags of a given image, so `docker push foo/bar` would push (e.g.) all of  `foo/bar:latest`, `foo:/bar:v1`, `foo/bar:v1.0.0`.
>
> Pushing all tags of an image was not desirable in many case, so docker v0.10.0 enhanced `docker push` to optionally specify a tag to push (`docker push foo/bar:v1`) (see https://github.com/moby/moby/issues/3411 and the pull request that implemented this: https://github.com/moby/moby/pull/4948).

I should mention that the changes I made in https://github.com/moby/moby/pull/40302 were in the _client_ (so that the CLI behavior matched the behavior of `docker pull`), mostly because we're trying to be conservative when making changes to the API (especially when changing behavior for API endpoints that have been in use for a _long_ time).

That means that at the _API_ level both currently still have `--all-tags` as a default (which is the problem you were hitting).


Having said the above, I (personally) think that we should consider changing the default, but make the behavior depend on API version, so that old API versions remain compatible.

Something like:

- Introduce an `all-tags` query parameter on the `POST /images/create` and `POST /images/{name}/push`
    - On API >= 1.42, If the `all-tags` parameter is not set, `latest` is assumed
    - On API >= 1.42, if _both_ the `all-tags` parameter _and_ the `tag` parameter are set, an error should be produced (conflicting options); a similar validation is currently handled client-side; https://github.com/moby/moby/pull/40302/files#r356040886
    - On API < 1.42, the old behavior is kept, which means that the `all-tags` query parameter is ignored, and is the default if no `tag` parameter is set
- Where needed, update the client and CLI to set the correct options, based on API version that's used (this may not actually be needed, as CLI `--all-tags` can both "remove" the `tag` parameter _and_ set the `all-tags`, which would make both "old" and "new" API do the right thing.
- Double check / verify changes for both the "pull" and "import" cases (`POST /images/create` is used both for pulling images and for importing images from a tar
- (orthogonal) perhaps we should consider having separate endpoints for "pull" and "import"


Note that this change would be a "breaking" change for older clients that don't use the versioned API endpoints, which, although deprecated (https://github.com/moby/moby/pull/28208), might still be common. I expect the use of (intentionally) "pulling" or "pushing" _all tags_ to be limited, but there may be some situations where this is used (considering third-party API clients, so outside of those using the `docker` CLI for this). Some options to ease this situation;

1. even though it's the "default", make sure the the client always sets the `tag` parameter explicitly.
    If an unversioned API call was made, and `tag` is not set, return a "warning" in the API response to inform the user that "latest" is pulled/pushed, and to push/pull all tags, the `all-tags` parameter should be used in API 1.42 and up
2. alternatively, make the `tag` parameter _required_ if `all-tags` is not set (should be either `:latest`,  `:tag`, or `@digest`).
    - **pro**: it's more clear on "intent" (`:latest` is mostly a convention; requiring it to be set makes it more clear what's meant).
    - **pro**: the breaking change won't be "silent": the API call will fail, and the user know that changes are needed in their client application.
    - **con**: makes the API somewhat more cumbersome to use (why set `?tag=latest` if it could be the default on the daemon side)

**Comment by tianon** [comment]
IMO your proposed plan of attack WRT introducing this break in API 1.42 makes sense, and my opinion on users who aren't using a versioned endpoint is that they should really fix that, even if they fix it by hard coding some specific version that meets their minimum supported Docker version :shrug:

It's more honest to the Docker daemon as to what their API expectations are, and less likely to be hit by ... breaking changes :sunglasses:

So in short, +1 on API 1.42 introducing `all-tags`, making it mutually exclusive with `tag`, and even making `tag` required if `all-tags` is not specified / `false` (makes it more clear to API consumers that something has changed and make it the client's responsibility to expand to `:latest` if it wants that to be implied like `docker pull` would -- explicit over implicit, _especially_ in an API that users don't often use by hand :smile:).

**Comment by tianon** [comment]
Coming back to this in the context of #44963, it would be really neat if we combined this change (explicit `all-tags`) with a change from `image`+`tag` to `ref` so that we could support push-by-digest on containerd. :eyes:

Maybe mutually exclusive `all-tags: repo/name` vs `ref: repo/name@sha256:xxx` / `ref: repo/name:foo` ?  I guess a more "backwards compatible" change would be to allow the existing `image` to reference a fully-qualified name (thus ignoring `tag` and going back to `all-tags` being a boolean).

**Comment by thaJeztah** [comment]
Note to self (as this came up in a discussion); `docker save` / `docker image save` also has this behavior; omitting `:latest` means "all tags"; 

It's documented, but I think it's "surprising behavior" for most users, as in most other places "no tag" means :latest so I think we should consider making a breaking change and fix that; https://docs.docker.com/reference/cli/docker/image/save/


> Produces a tarred repository to the standard output stream. Contains all parent layers, and all tags + versions, or specified repo:tag, for each argument provided.
